src/core/powerSave.cpp

```diff
@@ -1,64 +1,193 @@
+// powerSave.cpp
 #include "powerSave.h"
-#include "display.h"
+#include "led_control.h"
 #include "settings.h"
 
-/* Check if it's time to put the device to sleep */
+// How long to wait after dimming before full sleep
 #define SCREEN_OFF_DELAY 5000
+// Fade step delay in ms
+#define FADE_STEP_DELAY 5
+
+// --- Non‑blocking fade state ---
+static bool isFading = false;
+static unsigned long fadeStartTime = 0;
+static int fadeStartDisp, fadeStartLed;
+static int fadeEndDisp, fadeEndLed;
+static int fadeDuration;
+static bool fadingIn;
+
+// Forward declaration
+static void fadeAll(int startDisp, int startLed, int endDisp, int endLed);
+
+/**
+ * @brief Call this every loop (or task) to step an in‑progress fade.
+ */
+void updatePowerSave() {
+    if (!isFading) return;
+
+    unsigned long elapsed = millis() - fadeStartTime;
+    if (elapsed >= (unsigned long)fadeDuration) {
+        // final state
+        setBrightness(fadeEndDisp, false);
+        setLedBrightness(fadeEndLed);
```

Comment:
You will need to make the LED fading only happen on devices with LEDs, this is one place.

```
#ifdef HAS_RGB_LED
...
#endif
```